### PR TITLE
feat: Add basic TypeScript types

### PR DESCRIPTION
**Pull request by quinnturner**
## Description of the change

I do not expect these to be completely accurate.
However, they should be a decent starting point for further discussion!

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Maintenance
- [ ] New release

## Related issues

> Shortcut stories and GitHub issues (delete irrelevant)

- #7 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers assigned
- [ ] Issue from task tracker has a link to this pull request
- [ ] Changes have been reviewed by at least one other engineer

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  context?: string;
+}> {}
+export function useRollbar(): Rollbar;
+export function useRollbarConfiguration(config: Rollbar.Configuration): void;
+export function useRollbarContext(ctx?: string, isLayout?: boolean): void;
+export function useRollbarPerson(person: any): void;
+export function useRollbarCaptureEvent(metadata: object, level?: LEVEL): void;
+export function isValidLevel(): boolean;
+
+export function historyContext(
```
Comment (quinnturner) [inline index.d.ts:42]: I am unsure about most of `historyContext`, especially `filter`, `formatter`, and `v4x`s

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  | typeof LEVEL_INFO
+  | typeof LEVEL_WARN
+  | typeof LEVEL_ERROR
+  | typeof LEVEL_CRITICAL;
+
+export interface ErrorBoundaryProps {
+  children: ReactNode;
+  fallbackUI?: ReactNode
+  errorMessage?: string | (() => string)
+  extra?: Record<string | number, unknown> | (() => Record<string | number, unknown>)
```
Comment (quinnturner) [inline index.d.ts:20]: I am unsure about `extra`

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  level?: LEVEL | (() => LEVEL)
+  callback?: () => void
```
Comment (quinnturner) [inline index.d.ts:22]: I am unsure if callback has props

Comment (quinnturner) [reply index.d.ts]: Resolved by reusing Rollbar's types.

File: index.d.ts (hunk 1 above)
Comment (quinnturner) [reply index.d.ts:49]: Improved them but still not 100%

**Comment by calvinf** [comment]
It would be very helpful to have types as part of the module.

I found one component missing a type when I tried to use this myself - `Provider`.

Here's an implementation that worked for me:

```
type ProviderPropTypes = {
    Rollbar?: Rollbar
    config: Rollbar.Configuration
    instance?: Rollbar
  }

  export class Provider extends Component<ProviderPropTypes> {}
```

For anyone hacking this into their own project manually, you'll want to wrap the index.d.ts from this PR:

```typescript
declare module '@rollbar/react' {
 // types go here
}
```

And put it in a place that the types are picked up, e.g. if you have typeRoots in your `tsconfig.json`

```json
{
  "typeRoots": ["./types/"]
}
```

Then put the file as `rollbar-react.d.ts` in the `types` folder.

**Review by waltjones — APPROVED** [review]
@quinnturner thank you.

Merging this, and welcoming follow up PRs to fix any remaining issues.

**Comment by quinnturner** [comment]
@waltjones @calvinf, I added types for `Provider` here: https://github.com/rollbar/rollbar-react/pull/27

**Comment by waltjones** [comment]
@quinnturner @calvinf v0.9.0 is released now with both PRs included.

**Pull request merged**